core/trino-spi/src/main/java/io/trino/spi/exchange/ExchangeSourceOutputSelector.java

```diff
@@ -0,0 +1,279 @@
+        private boolean finalSelector;
+
+        public Builder include(ExchangeId exchangeId, int taskPartitionId, int attemptId)
+        {
+            values.computeIfAbsent(exchangeId, key -> new ValuesBuilder()).include(taskPartitionId, attemptId);
+            return this;
+        }
+
+        public Builder exclude(ExchangeId exchangeId, int taskPartitionId)
+        {
+            values.computeIfAbsent(exchangeId, key -> new ValuesBuilder()).exclude(taskPartitionId);
+            return this;
+        }
+
+        public Builder setFinal()
+        {
+            this.finalSelector = true;
+            return this;
+        }
+
+        public ExchangeSourceOutputSelector build()
+        {
+            return new ExchangeSourceOutputSelector(
+                    nextVersion++,
+                    values.entrySet().stream()
+                            .collect(toMap(Map.Entry::getKey, entry -> entry.getValue().build(finalSelector))),
+                    finalSelector);
+        }
+    }
+
+    private static class ValuesBuilder
+    {
+        private Slice values = Slices.allocate(0);
+        private int maxTaskPartitionId;
+
+        public void include(int taskPartitionId, int attemptId)
+        {
+            updateMaxTaskPartitionIdAndEnsureCapacity(taskPartitionId);
+            if (attemptId < 0 || attemptId > Byte.MAX_VALUE) {
+                throw new IllegalArgumentException("unexpected attemptId: " + attemptId);
+            }
+            values.setByte(taskPartitionId, (byte) attemptId);
+        }
+
+        public void exclude(int taskPartitionId)
+        {
+            updateMaxTaskPartitionIdAndEnsureCapacity(taskPartitionId);
+            values.setByte(taskPartitionId, EXCLUDED.getValue());
+        }
+
+        private void updateMaxTaskPartitionIdAndEnsureCapacity(int taskPartitionId)
+        {
+            if (taskPartitionId > maxTaskPartitionId) {
+                maxTaskPartitionId = taskPartitionId;
+            }
+            if (taskPartitionId < values.length()) {
+                return;
+            }
+            byte[] newValues = new byte[(maxTaskPartitionId + 1) * 2];
+            fill(newValues, UNKNOWN.getValue());
+            values.getBytes(0, newValues, 0, values.length());
+            values = Slices.wrappedBuffer(newValues);
+        }
+
+        public Slice build(boolean finalSelector)
+        {
+            int partitionCount = maxTaskPartitionId + 1;
+            Slice result = Slices.allocate(partitionCount);
+            result.setBytes(0, values, 0, partitionCount);
+            if (finalSelector) {
+                for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
+                    byte selectedAttempt = result.getByte(partitionId);
+                    if (selectedAttempt == UNKNOWN.getValue()) {
+                        throw new IllegalStateException("Attempt is unknow for partition: " + partitionId);
```

Comment:
unknown

Also you should probably provide totalNumberOfPartitions and verify that `partitionCount` is equal to that if you are building final selector (or do you validate it somewhere?)